### PR TITLE
[20.09] Fix Anndata index access, use numpy-safe JSON encoder

### PR DESCRIPTION
**Pull request by mtekman**
Re-added try/except block due to failing datasets: https://usegalaxy.eu/u/pmoreno/h/test-older-and-newer-anndata-from-zero

references issue here https://github.com/galaxyproject/galaxy/issues/10639

### CODE REVIEW

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -981,78 +981,81 @@ def sniff(self, filename):
...
+                        size = int(tmp.size)
+                    else:
+                        layers = [util.unicodify(x) for x in list(tmp.keys())]
+                        count = len(layers)
+                        size = lennames
+                    return (layers, count, size)
+
+                if 'obs' in dataset.metadata.layers_names:
+                    tmp = anndata_file["obs"]
+                    dataset.metadata.obs_names = [util.unicodify(x) for x in tmp["index"]]
```
Comment (pcm32) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
                    tmp = anndata_file["obs"]
                    if "index" in tmp:
                        obs_index = "index"
                    else:
                        obs_index = "_index"
                    dataset.metadata.obs_names = [util.unicodify(x) for x in tmp[obs_index]]
```
Maybe we can take the chance to cater for the "_index" 0.7.4 case

**Comment by bernt-matthias** [comment]
Seems preferable to add a test case for the failing case and then really fix the problem. If the guess in the linked issue is correct one would need to access `_index` instead of `index` of the later is unavailable.

**Comment by pcm32** [comment]
Sure, we can leave that fix for later and at least avoid failing now with the exception handling. I have passed @mtekman  some test data, maybe he intended to do it on another PR and I rushed. Anyway is fine for me.

File: lib/galaxy/datatypes/binary.py (hunk 1 above)
Comment (pcm32) [reply lib/galaxy/datatypes/binary.py]: feel free to neglect as per below comments.

**Comment by mtekman** [comment]
@bernt This is just a hotfix for now

> Seems preferable to add a test case for the failing case and then really fix the problem. If the guess in the linked issue is correct one would need to access `_index` instead of `index` of the later is unavailable.

this is just a hotfix for now to get the datasets running before next week

Another PR will try address the underlying issue

**Comment by pcm32** [comment]
maybe it would be better then to revert my suggestion and keep this clean for just dealing with the exception/error.

**Comment by mtekman** [comment]
I see no harm in trying to parse both types of datasets, and catching the outliers at the same time

**Comment by pcm32** [comment]
ok, code above is not tested though, but should be fine.

**Comment by pcm32** [comment]
we might want to repeat the operation then on var, where apparently the same applies.

**Comment by mtekman** [comment]
oh good point I hadn't considered that

**Comment by mvdbeek** [comment]
Can we please fix this properly ? I am not a fan of the try/except ... either the information that `set_meta` generates is important, and then we should parse it correctly, or it is not, in which case we should not try to parse it.

If the format is in flux entirely maybe you'd want to raise this with the authors. If I look at https://github.com/theislab/anndata/blame/77a633cc825b428d02dabdf75491841c3a175a8d/anndata/_io/h5ad.py#L251 I would think that the index name is stored in `_index`, always, so what's in here doesn't look correct either.

If we generate wrong metadata that's worse then erroring.

**Comment by pcm32** [comment]
Unfortunately index is not always on "_index", and people could upload AnnData that are older.

**Comment by mtekman** [comment]
@mvdbeek These datasets are likely to further evolve rather than stabilise, and I can't think of a future-proof solution other than try/except

**Comment by mvdbeek** [comment]
> @mvdbeek These datasets are likely to further evolve rather than stabilise, and I can't think of a future-proof solution other than try/except

but then you'll never know that something is not as expected with a new dataset. Isn't there some kind of version indication if this changes so rapidly ?

**Comment by mtekman** [comment]
Likely. I did previously search for version tags in the test data that we have, but was unable to find anything...

I guess we could remove the try/except and test for a few required keys as you say, and then still return a green dataset if none are found

**Comment by pcm32** [comment]
> I guess we could remove the try/except and test for a few required keys as you say, and then still return a green dataset if none are found

Yes please, the current state is just blocking any further tool execution with the red dataset. Maybe if the index key is not found we could have a message saying this, but avoid leaving the dataset in failure state?

**Comment by pcm32** [comment]
> If I look at https://github.com/theislab/anndata/blame/77a633cc825b428d02dabdf75491841c3a175a8d/anndata/_io/h5ad.py#L251 I would think that the index name is stored in `_index`, always, so what's in here doesn't look correct either.

In the associated issue #10640 there is a link to AnnData code showing that even in the newer version there are circumstances where "index" instead of "_index" would be used. This is partly why this flew under the radar.

**Comment by mvdbeek** [comment]
> Maybe if the index key is not found we could have a message saying this, but avoid leaving the dataset in failure state?

Does the PR in its current form, minus the try/except, fail any known datasets ?
If not, it seems much preferable to remove them, so we actually know when there's a weird dataset appearing.

Keep in mind that faulty metadata can lead to wrong analysis results. That is much worse than having to report a bug.

**Comment by mtekman** [comment]
> Does the PR in its current form, minus the try/except, fail any known datasets ?

We need more datasets, and we have some, but they need to be greatly reduced in size before we can add them as test data to galaxy

**Comment by mvdbeek** [comment]
The test can't fail now, can you push the changes ?

**Comment by mtekman** [comment]
> The test can't fail now, can you push the changes ?

Some of the newer datasets are failing locally... give me 30 minutes to debug and then I'll push

**Comment by mtekman** [comment]
I was unable to debug why things were failing locally, so I pushed in the hopes that you guys might....

File: lib/galaxy/model/metadata.py [hunk 2]
```diff
@@ -212,9 +211,9 @@ def to_JSON_dict(self, filename=None):
         if '__validated_state_message__' in dataset_meta_dict:
             meta_dict['__validated_state_message__'] = dataset_meta_dict['__validated_state_message__']
         if filename is None:
```
Comment (nsoranzo) [inline lib/galaxy/model/metadata.py:214]:
```suggestion
        encoded_meta_dict = galaxy.model.custom_types.json_encoder.encode(meta_dict)
        if filename is None:
```

File: lib/galaxy/model/metadata.py [hunk 3]
```diff
@@ -212,9 +211,9 @@ def to_JSON_dict(self, filename=None):
...
-            return json.dumps(meta_dict)
+            return galaxy.model.custom_types.json_encoder.encode(meta_dict)
```
Comment (nsoranzo) [inline lib/galaxy/model/metadata.py]:
```suggestion
            return encoded_meta_dict
```

File: lib/galaxy/model/metadata.py [hunk 4]
```diff
@@ -212,9 +211,9 @@ def to_JSON_dict(self, filename=None):
...
         with open(filename, 'wt+') as fh:
-            json.dump(meta_dict, fh)
+            fh.write(galaxy.model.custom_types.json_encoder.encode(meta_dict))
```
Comment (nsoranzo) [inline lib/galaxy/model/metadata.py]:
```suggestion
            fh.write(encoded_meta_dict)
```

File: lib/galaxy/datatypes/binary.py [hunk 5]
```diff
@@ -998,21 +1010,29 @@ def set_meta(self, dataset, overwrite=True, **kwd):
...
                     size = lennames
                 return (layers, count, size)
 
             if 'obs' in dataset.metadata.layers_names:
                 tmp = anndata_file["obs"]
-                dataset.metadata.obs_names = [util.unicodify(x) for x in tmp["index"]]
-                dataset.metadata.obs_layers, \
-                    dataset.metadata.obs_count, \
-                    dataset.metadata.obs_size = _layercountsize(tmp, len(dataset.metadata.obs_names))
+                obs_index = False
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
                obs_index = None
```

File: lib/galaxy/datatypes/binary.py [hunk 6]
```diff
@@ -1022,17 +1042,25 @@ def _layercountsize(tmp, lennames=0):
...
+                dataset.metadata.raw_var_size = z
 
             if 'var' in dataset.metadata.layers_names:
                 tmp = anndata_file["var"]
-                # row names are never used in preview windows
-                # dataset.metadata.var_names = tmp["index"]
-                dataset.metadata.var_layers, \
-                    dataset.metadata.var_count, \
-                    dataset.metadata.var_size = _layercountsize(tmp, len(tmp["index"]))
+                var_index = False
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
                var_index = None
```

**Review by mvdbeek — APPROVED** [review]
Thank you @mtekman & @pcm32 & @nsoranzo !

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**